### PR TITLE
Support building raspidistro VLC (with mmal enabled)

### PR DESCRIPTION
**Pull request by EasyIP2023**
Hello all,

Opening this pull request so it's easier to track progress.

This PR is connected to
* https://github.com/agherzan/meta-raspberrypi/issues/978
* https://github.com/agherzan/meta-raspberrypi/issues/622

**The raspidistro-vlc recipe is still a working progress on branch master/honister.**

**What I did**
1. Copied VLC + ffmpeg recipes that can be found online and set them up to clone
     * https://github.com/RPi-Distro/ffmpeg
     * https://github.com/RPi-Distro/vlc
2. Took the patches found in both repos ``debian/patches`` directory and applied them
     * **Note:** The critical VLC patch https://github.com/EasyIP2023/meta-raspberrypi/blob/hw-mmal-vlc/recipes-multimedia/rpidistro-vlc/files/0006-Use-packageconfig-to-detect-mmal-support.patch. Uses pkg-config to retrieve necessary compile flags as @kraj did in https://github.com/openembedded/meta-openembedded/blob/master/meta-multimedia/recipes-multimedia/vlc/vlc/0004-Use-packageconfig-to-detect-mmal-support.patch with added twist of checking for ``vc_tv_unregister_callback_full``. Unsigned as I wasn't sure if I should.
     * Applying patches gives access to the ``./configure --enable-mmal-avcodec`` flag which is not available in upstream VLC.
3. Update PACKAGECONFIG to include configure flags added after patches were applied

**How I did it**

Cloned repos bellow
* YP Core                        (honister 3.4.1) : commit 4ce984316d038eaed2f48533b31d0a1898942f0a
* meta-openembbeded (master/honister) : commit 5e0f1edc96cf2206cb9333dcab7acdd634813f6f
* meta-raspberrypi        (master/honister) : commit 58cc6625395c251e68604de587c332f07a6acd3b

Added bellow inside ``local.conf``
```
MACHINE = "raspberrypi4"
EXTRA_IMAGE_FEATURES:append = " x11 opengl"
LICENSE_FLAGS_WHITELIST = "commercial_ffmpeg commercial_vlc commercial_rpidistro-ffmpeg commercial_rpidistro-vlc"
LICENSE_FLAGS_WHITELIST:append = " commercial_x264 commercial_faad2 commercial_mpeg2dec"
IMAGE_INSTALL:append = " rpidistro-vlc"

ENABLE_UART = "1"
DISABLE_VC4GRAPHICS = "1"
```

Modified mesa-gl recipe source so that libgbm could build without dri support. Mainly to get pasted ffmpeg compile.

With personal setup bellow are the results
* YP Core          (Dunfell 3.1.10)  : commit 2a848e95074318f3a243df7b3f40513a13173a82
* meta-openembedded (Dunfell) : commit 11eae114522a6befa06c7f4021a83bc016133543
* meta-raspberrypi        (Dunfell) : commit 934064a01903b2ba9a82be93b3f0efdb4543a0e8
all-flags

### CODE REVIEW

File: recipes-multimedia/rpidistro-vlc/rpidistro-vlc_3.0.12.bb [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  file://0009-fix-EGL-macro-undeclared-and-EGLImageKHR.patch \
+  file://0010-fix-numeric_limits-is-not-a-member-of-std.patch \
+"
+
+SRCREV = "f7fd69f12a3b89d03768fa3bd468e8f33cd1dc7c"
+
+S = "${WORKDIR}/git"
+
+# While this item does not require it, it depends on ffmpeg which does
+LICENSE_FLAGS = "commercial"
```
Comment (agherzan) [inline recipes-multimedia/rpidistro-vlc/rpidistro-vlc_3.0.12.bb:27]: If so, Shouldn't ffmpeg have it?

File: recipes-multimedia/rpidistro-vlc/files/0001-configure-fix-linking-on-RISC-V-ISA.patch [hunk 2]
```diff
@@ -0,0 +1,18 @@
+From: =?utf-8?q?R=C3=A9mi_Denis-Courmont?= <remi@remlab.net>
+Date: Sat, 16 Jun 2018 21:31:45 +0300
+Subject: configure: fix linking on RISC-V ISA
```
Comment (agherzan) [inline recipes-multimedia/rpidistro-vlc/files/0001-configure-fix-linking-on-RISC-V-ISA.patch:3]: Upstream status missing for all these patches.

File: recipes-multimedia/rpidistro-ffmpeg/files/0001-avcodec-arm-sbcenc-avoid-callee-preserved-vfp-regist.patch [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+the following way:
+ ff_sbc_analyze_4_neon: q2-q5 => q8-q11, then q1-q4 => q8-q11
+ ff_sbc_analyze_8_neon: q2-q9 => q8-q15
+
+The reason for using these replacements is to keep closely related
+sets of registers consecutively numbered which hopefully makes the
+code more easy to follow. Since this commit only reallocates
+registers, it should have no performance impact.
+
+Signed-off-by: James Cowgill <jcowgill@debian.org>
```
Comment (agherzan) [inline recipes-multimedia/rpidistro-ffmpeg/files/0001-avcodec-arm-sbcenc-avoid-callee-preserved-vfp-regist.patch:22]: Upstream status is missing.

File: recipes-multimedia/rpidistro-ffmpeg/rpidistro-ffmpeg_4.3.2.bb [hunk 4]
```diff
@@ -0,0 +1,181 @@
+
```
Comment (agherzan) [inline recipes-multimedia/rpidistro-ffmpeg/rpidistro-ffmpeg_4.3.2.bb:1]: Superfluous new line.

**Review by agherzan — CHANGES_REQUESTED** [review]
Looks good generally. Some small comments attached.

File: recipes-multimedia/rpidistro-ffmpeg/rpidistro-ffmpeg_4.3.2.bb (hunk 4 above)
Comment (EasyIP2023) [reply recipes-multimedia/rpidistro-ffmpeg/rpidistro-ffmpeg_4.3.2.bb]: fixed https://github.com/agherzan/meta-raspberrypi/pull/981/commits/a4890a4e8ccbe8b89db3a7b7f17943c37e561b01

**Comment by EasyIP2023** [comment]
Hey @agherzan

Got a full build going to check. Do you think it would be better to override the ``do_patch`` function as most of the patches were taken from the github repos ``debian/patches`` folder versus me copying them directly. Not sure I can modify the corresponding upstream status in said patch headers. I can for my patches though.

File: recipes-multimedia/rpidistro-vlc/rpidistro-vlc_3.0.12.bb (hunk 1 above)
Comment (EasyIP2023) [reply recipes-multimedia/rpidistro-vlc/rpidistro-vlc_3.0.12.bb:27]: The rpidistro-ffmpeg recipe should also have it

**Comment by EasyIP2023** [comment]
Hey @agherzan

Nevermind haha. Didn't know I could just modify the patch header directly and have them still apply. Just learned something new. Just added the upstream status in each patch. According to https://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines they all should be pending

File: recipes-multimedia/rpidistro-vlc/files/0001-configure-fix-linking-on-RISC-V-ISA.patch (hunk 2 above)
Comment (EasyIP2023) [reply recipes-multimedia/rpidistro-vlc/files/0001-configure-fix-linking-on-RISC-V-ISA.patch:3]: fixed: https://github.com/agherzan/meta-raspberrypi/pull/981/commits/728905704d518db38babc2bed66811abd3673284

File: recipes-multimedia/rpidistro-vlc/rpidistro-vlc_3.0.12.bb (hunk 1 above)
Comment (agherzan) [reply recipes-multimedia/rpidistro-vlc/rpidistro-vlc_3.0.12.bb:27]: Sure. But should it also be added to all the recipes that depend on one that is commercially flagged? Why is that the case?

File: recipes-multimedia/rpidistro-ffmpeg/files/0001-avcodec-arm-sbcenc-avoid-callee-preserved-vfp-regist.patch (hunk 3 above)
Comment (agherzan) [reply recipes-multimedia/rpidistro-ffmpeg/files/0001-avcodec-arm-sbcenc-avoid-callee-preserved-vfp-regist.patch:24]: This seems to have been missed.

**Comment by agherzan** [comment]
@EasyIP2023 All good. They are either pending or backports I guess.

File: conf/machine/include/rpi-default-providers.inc [hunk 5]
```diff
@@ -7,6 +7,8 @@ PREFERRED_PROVIDER_virtual/libgles2 ?= "${@bb.utils.contains("MACHINE_FEATURES",
 PREFERRED_PROVIDER_virtual/libgl ?= "${@bb.utils.contains("MACHINE_FEATURES", "vc4graphics", "mesa", "mesa-gl", d)}"
 PREFERRED_PROVIDER_virtual/mesa ?= "${@bb.utils.contains("MACHINE_FEATURES", "vc4graphics", "mesa", "mesa-gl", d)}"
 PREFERRED_PROVIDER_virtual/libgbm ?= "${@bb.utils.contains("MACHINE_FEATURES", "vc4graphics", "mesa", "mesa-gl", d)}"
+PREFERRED_PROVIDER_virtual/vlc ?= "vlc"
```
Comment (agherzan) [inline conf/machine/include/rpi-default-providers.inc:10]: I like this. We should document it somewhere though - see documentation files.

**Review by agherzan — CHANGES_REQUESTED** [review]
We are getting closer. Can we also squash and clean up the history? There is no reason to have that set of iterations when we introduce new recipes/mechanisms.

**Comment by EasyIP2023** [comment]
Hey,

@agherzan accidentally deleted my branch when pushing picked + squashed commits. It's pushed back up again. Can we reopen or do I need to create another pull request?

Changes include
  * ``ENABLE_RPI_DISTRO_VLC = "1"`` variable for adding rpidistro-vlc packages into the image
  * removal of the ``LICENSE_FLAGS`` in the raspidistro-vlc recipe as it is not needed
  * Cleaned up git history

**Pull request closed without merging**